examples/knative/powercli/kn-pcli-tag/handler.ps1

```diff
@@ -0,0 +1,64 @@
+Function Process-Handler {
+   param(
+      [Parameter(Position=0,Mandatory=$true)][CloudNative.CloudEvents.CloudEvent]$CloudEvent
+   )
+
+   # Decode CloudEvent
+   $cloudEventData = $cloudEvent | Read-CloudEventJsonData -ErrorAction SilentlyContinue -Depth 10
+   if($cloudEventData -eq $null) {
+      $cloudEventData = $cloudEvent | Read-CloudEventData
+   }
+
+   if(${env:FUNCTION_DEBUG} -eq "true") {
+      Write-Host "DEBUG: K8s Secrets:`n${env:TAG_SECRET}`n"
+
+      Write-Host "DEBUG: CloudEventData`n $(${cloudEventData} | Out-String)`n"
+   }
+
+   if(${env:TAG_SECRET}) {
+      $jsonSecrets = ${env:TAG_SECRET} | ConvertFrom-Json
+   } else {
+      Write-Host "K8s secrets `$env:TAG does not look to be defined"
+      break
+   }
+
+   # Extract all tag secrets for ease of use in function
+   $VCENTER_SERVER = ${jsonSecrets}.VCENTER_SERVER
+   $VCENTER_USERNAME = ${jsonSecrets}.VCENTER_USERNAME
+   $VCENTER_PASSWORD = ${jsonSecrets}.VCENTER_PASSWORD
+   $VCENTER_TAG_NAME = ${jsonSecrets}.VCENTER_TAG_NAME
+
+   # Extract VM Name from event
+   $vmName = $cloudEventData.Vm.Name
+
+   # Configure TLS 1.2/1.3 support as this is required for latest vSphere release
+   [System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor [System.Net.SecurityProtocolType]::Tls12 -bor [System.Net.SecurityProtocolType]::Tls13
+
+   Write-Host "$(Get-Date) - Configuring PowerCLI Configuration Settings`n"
+   try {
+      Set-PowerCLIConfiguration -InvalidCertificateAction:Ignore -ParticipateInCeip:$true -Confirm:$false -WarningVariable Warn | Out-Null
+   } catch {
+      Write-Host "Failed to run Set-PowerCLIConfiguration"
+   }
+
+   Write-Host "$(Get-Date) - Connecting to vCenter Server $VCENTER_SERVER`n"
+   try {
+      Connect-VIServer -Server $VCENTER_SERVER -User $VCENTER_USERNAME -Password $VCENTER_PASSWORD
+   } catch {
+      Write-Host "Failed to run Connect-VIServer"
+   }
+
+   Write-Host "$(Get-Date) - Applying vSphere Tag `"$VCENTER_TAG_NAME`" to $vmName ...`n"
+   try {
+      Get-VM $vmName | New-TagAssignment -Tag (Get-Tag -Name $VCENTER_TAG_NAME) -Confirm:$false
```

Comment:
IIUC, get-tag performs another roundtrip to retrieve its UUID. Could be avoided when using UUID (at least that's what we did in other examples) to speed up performance and avoid vc roundtrips.

Reply:
Correct. However, this won't be possible if we're using high-level PowerCLI cmdlets which retrieves objects by name. If we wanted to do the same thing in Python or other SDK, which doesn't provide the friendly name handling, then we'd have to call the vSphere Tagging API directly, which most users will probably not do due to ease of use of the high level cmdlets